src/LagrangianParticleTracking/update_particle_properties.jl

```diff
@@ -157,7 +157,7 @@ function update_particle_properties!(lagrangian_particles, model, Δt)
     advect_particles_kernel! = _advect_particles!(device(arch), workgroup, worksize)
 
     advect_particles_event = advect_particles_kernel!(lagrangian_particles.properties, lagrangian_particles.restitution, model.grid, Δt,
-                                                      datatuple(model.velocities),
+                                                      total_velocities(model),
```

Comment:
Why not pass in all the velocity fields explicitly?

Reply:
As in pass velocities and background fields separately, or u/v/w separately? 

This was basically just the minimum change from what is currently implemented.

Reply:
The answer might be that different models have different "total velocities", therefore it's simpler to dispatch at this level than farther out?

Generally when we make a change I think it's worth discussing whether the best choice is (1) extending the existing design or (2) implementing a new design (since the original code may not have been written with this concept in mind). Here `total_velocities` increase the burden for implementing new `AbstractModel` --- eg do we want to encode an expectation that all `AbstractModel` have the notion of "velocities"? This _could_ make sense as a design choice that all models wishing to participate in Lagrangian particle advection need to define `total_velocities` (or perhaps just "velocities"?) It's worth discussing.

Reply:
Yeah that's why I had to implement it like this because hydrostatic models don't have background velocities like nonhydrostatic models. 

I see what you mean. While it's only being used here having the expectation that a model will have velocities if it's going to have particles seems reasonable? Up until now it's just been assumed that at `model.velocities` it would find a `u`, `v`, and `w` velocity components too (which I assume could be why shallow water models don't support particles).

Another thing to consider if we're coming up with a broader notion of models having velocities is that this doesn't include biogeochemical drift velocities so it more a specific momentum.

Reply:
one challenge is that I think as a user I would expect `total_velocities` to return an `AbstractField` or something that can be used in operations. But with `total_velocities` sometimes we get a `NamedTuple` of fields, sometimes we get a `NamedTuple` of `SumOfArrays`...

Reply:
Another solution would be to define an `advect_particles!(lagrangian_particles, model, Δt)` function that dispatches on the type of `model` ?

Reply:
Then each model decides how to advect particles (without having to make broader decisions on what constitutes "total velocities"

Reply:
Noticing also that this module is strange in that it defines a function of `model` before models even exist.

@jagoosw I think we should refactor this particle advection code to make more sense and avoid `total_velocities`. We can do that in this PR or another. If you want to tackle it here, can I commit to this PR?

Reply:
A consistent source of paranoia is how difficult it is going to be to split up this repo... we will very likely need to do that at some point, so I'm worried about circular dependencies and such (basically, all the functions that have to be defined at top-level --- thats bad / worrisome), which could make that difficult.

Reply:
Okay yeah this makes sense and I like that idea. Happy to tackle it here and for you to commit to this PR

Reply:
I believe that particles are defined before the models so that the type of particles can be forced to be `Union{Nothing, LagrangianParticles}` which probably isn't needed?

Reply:
I opened an issue to make this concern more clear: https://github.com/CliMA/Oceananigans.jl/issues/3048

Reply:
Yeah this seems like a good idea. Having a generic type means users could create their own things that go in the particles slot instead (which I think could make the interface in OceanBioME nicer).

Reply:
I've set this up now, but not sure how to deal with the total velocities still? I guess I could define the function `total_velocities` in the `LagrangianParticles` dispatched on different model types now that it is after the models? And then its just a local funciton for the particles, but does mean more modifcaiton for usign particles in a different mode?

Reply:
I have some ideas for how this should be implemented

Reply:
Been in a few meetings today, but I can have a chance at coding in a bit